src/hotspot/share/jvmci/jvmciRuntime.cpp

```diff
@@ -904,19 +958,232 @@ static void _flush_log() {
+    new_runtime->attach_thread(thread);
+  }
+}
+
+bool JVMCIRuntime::detach_thread(JavaThread* thread, const char* reason, bool can_destroy_javavm) {
+  if (this == JVMCI::_shutdown_compiler_runtime || JVMCI::in_shutdown()) {
+    // Do minimal work when shutting down JVMCI
+    thread->set_libjvmci_runtime(NULL);
+    return false;
+  }
+  bool should_shutdown;
+  bool destroyed_javavm = false;
+  {
+    MutexLocker locker(JVMCI_lock);
+    _num_attached_threads--;
+    JVMCI_event_1("detaching from JVMCI runtime %d: %s (%d other threads still attached)", _id, reason, _num_attached_threads);
+    should_shutdown = _num_attached_threads == 0 && !JVMCI::in_shutdown();
+    if (should_shutdown && !can_destroy_javavm) {
+      // If it's not possible to destroy the JavaVM on this thread then the VM must
+      // not be shutdown. This can happen when a shared library thread is the last
+      // thread to detach from a shared library JavaVM (e.g. GraalServiceThread).
+      JVMCI_event_1("Cancelled shut down of JVMCI runtime %d", _id);
+      should_shutdown = false;
+    }
+    if (should_shutdown) {
+      // Prevent other threads from attaching to this runtime
+      // while it is shutting down and destroying its JavaVM
+      _num_attached_threads = cannot_be_attached;
+    }
+  }
+  if (should_shutdown) {
+    // Release the JavaVM resources associated with this
+    // runtime once there are no threads attached to it.
+    shutdown();
+    if (can_destroy_javavm) {
+      destroyed_javavm = destroy_shared_library_javavm();
+      if (destroyed_javavm) {
+        // Can release all handles now that there's no code executing
+        // that could be using them. Handles for the Java JVMCI runtime
+        // are never released as we cannot guarantee all compiler threads
+        // using it have been stopped.
+        int released = release_and_clear_globals();
+        JVMCI_event_1("releasing handles for JVMCI runtime %d: oop handles=%d, metadata handles={total=%d, live=%d, blocks=%d}",
+            _id,
+            released,
+            _metadata_handles->num_handles(),
+            _metadata_handles->num_handles() - _metadata_handles->num_free_handles(),
+            _metadata_handles->num_blocks());
+
+        // No need to acquire _lock since this is the only thread accessing this runtime
+        _metadata_handles->clear();
+      }
+    }
+    // Allow other threads to attach to this runtime now
+    MutexLocker locker(JVMCI_lock);
+    _num_attached_threads = 0;
+    if (JVMCI::using_singleton_shared_library_runtime()) {
+      // Notify any thread waiting to attach to the
+      // singleton JVMCIRuntime
+      JVMCI_lock->notify();
+    }
+  }
+  thread->set_libjvmci_runtime(NULL);
+  JVMCI_event_1("detached from JVMCI runtime %d", _id);
+  return destroyed_javavm;
+}
+
+JNIEnv* JVMCIRuntime::init_shared_library_javavm() {
+  MutexLocker locker(_lock);
+  JavaVM* javaVM = (JavaVM*) _shared_library_javavm;
```

Comment:
Pointless cast.

Reply:
Without that cast, I get this compiler error:
```
/Users/dnsimon/jdk-jdk/open/src/hotspot/share/jvmci/jvmciRuntime.cpp:1191:11: error: cannot initialize a variable of type 'JavaVM *' (aka 'JavaVM_ *') with an lvalue of type 'volatile JavaVM *' (aka 'volatile JavaVM_ *')
  JavaVM* javaVM = _shared_library_javavm;
          ^        ~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
I forgot that `_shared_library_javavm` is declared `volatile JavaVM*`.  So what's needed is a const_cast to discard the volatile.  But why is it pointer to volatile?  I'm not seeing any reason for that; the out argument for `JNI_CreateJavaVM` doesn't specify volatile.  I wonder if it's supposed to instead be `JavaVM* volatile` to (attempt to, though maybe not correctly) support unlocked reads with concurrent locked writes?

Reply:
It is declared volatile for correct double checked locking in the code prior to this PR:
```
JNIEnv* JVMCIRuntime::init_shared_library_javavm() {
  JavaVM* javaVM = (JavaVM*) _shared_library_javavm;
  if (javaVM == NULL) {
    MutexLocker locker(JVMCI_lock);
    // Check again under JVMCI_lock
    javaVM = (JavaVM*) _shared_library_javavm;
    if (javaVM != NULL) {
      return NULL;
    }
```
However, now that `JVMCIRuntime::_lock` is acquired at the entry to `JVMCIRuntime::init_shared_library_javavm` there is indeed no longer a need for it to be volatile.

Reply:
The declaration was wrong for that usage - it should have been "volatile pointer to X" (`JavaVM* volatile`) but was declared "pointer to volatile X" (`volatile JavaVM*`).  That also wasn't a correct implementation of DCLP.  Good thing you are changing this.  So yes, the volatile should be nuked and this cast and the similar one in destroy_shared_library_javavm can be removed.

Reply:
Thanks for the correction. I have removed `volatile` altogether now. It's also a good reminder of the subtleties in C++ that do not exist in Java.

Reply:
Fixed: https://github.com/openjdk/jdk/pull/8262/commits/8aacbbaa2c3909694e83000ca0f8366fc6dc7931